runtime/protocol/aws-protocol-core/common/src/aws/smithy/kotlin/runtime/awsprotocol/AwsQueryCompatibleErrorDetails.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.awsprotocol
+
+import aws.smithy.kotlin.runtime.InternalApi
+import aws.smithy.kotlin.runtime.ServiceErrorMetadata
+import aws.smithy.kotlin.runtime.ServiceException
+
+/**
+ * Header name for awsQuery-compatible error values.
+ */
+@InternalApi
+public const val X_AMZN_QUERY_ERROR_HEADER: String = "x-amzn-query-error"
+
+/**
+ * Error details presented for backwards-compatibility by services that have migrated from awsQuery.
+ */
+@InternalApi
+public data class AwsQueryCompatibleErrorDetails(
+    public val code: String,
+    public val type: ServiceException.ErrorType,
+) {
+    public companion object {
+        public fun parse(value: String): AwsQueryCompatibleErrorDetails = parseImpl(value)
+    }
+}
+
+/**
+ * Set awsQuery error details on a [ServiceException]
+ */
+@InternalApi
+public fun setAwsQueryCompatibleErrorMetadata(exception: Any, error: AwsQueryCompatibleErrorDetails) {
```

Comment:
Is there a reason the parsing is split between `parse` and `parseImpl`?

Reply:
Nothing other than to avoid cluttering the main class definition.

We do this for `Url` as well, although that parsing implementation is far more complex.